wrapper/src/main/java/software/amazon/jdbc/MonitorPluginService.java

```diff
@@ -0,0 +1,795 @@
+
+    final Map<String, EnumSet<NodeChangeOptions>> changes = new HashMap<>();
+
+    for (final Map.Entry<String, HostSpec> entry : oldHostMap.entrySet()) {
+      final HostSpec correspondingNewHost = newHostMap.get(entry.getKey());
+      if (correspondingNewHost == null) {
+        // host deleted
+        changes.put(entry.getKey(), EnumSet.of(NodeChangeOptions.NODE_DELETED));
+      } else {
+        // host maybe changed
+        final EnumSet<NodeChangeOptions> hostChanges = compare(entry.getValue(), correspondingNewHost);
+        if (!hostChanges.isEmpty()) {
+          changes.put(entry.getKey(), hostChanges);
+        }
+      }
+    }
+
+    for (final Map.Entry<String, HostSpec> entry : newHostMap.entrySet()) {
+      if (!oldHostMap.containsKey(entry.getKey())) {
+        // host added
+        changes.put(entry.getKey(), EnumSet.of(NodeChangeOptions.NODE_ADDED));
+      }
+    }
+
+    if (!changes.isEmpty()) {
+      this.allHosts = newHosts != null ? newHosts : new ArrayList<>();
+      this.pluginManager.notifyNodeListChanged(changes);
+    }
+  }
+
+  @Override
+  public boolean isStaticHostListProvider() {
+    return this.getHostListProvider() instanceof StaticHostListProvider;
+  }
+
+  @Override
+  public void setHostListProvider(final HostListProvider hostListProvider) {
+    this.hostListProvider = hostListProvider;
+  }
+
+  @Override
+  public Connection connect(final HostSpec hostSpec, final Properties props) throws SQLException {
+    return this.connect(hostSpec, props, null);
+  }
+
+  @Override
+  public Connection connect(
+      final HostSpec hostSpec,
+      final Properties props,
+      final @Nullable ConnectionPlugin pluginToSkip)
+      throws SQLException {
+    return this.pluginManager.connect(
+        this.driverProtocol, hostSpec, props, this.currentConnection == null, pluginToSkip);
+  }
+
+  @Override
+  public Connection forceConnect(
+      final HostSpec hostSpec,
+      final Properties props)
+      throws SQLException {
+    return this.forceConnect(hostSpec, props, null);
+  }
+
+  @Override
+  public Connection forceConnect(
+      final HostSpec hostSpec,
+      final Properties props,
+      final @Nullable ConnectionPlugin pluginToSkip)
+      throws SQLException {
+    return this.pluginManager.forceConnect(
+        this.driverProtocol, hostSpec, props, this.currentConnection == null, pluginToSkip);
+  }
+
+  private void updateHostAvailability(final List<HostSpec> hosts) {
+    for (final HostSpec host : hosts) {
+      final HostAvailability availability = hostAvailabilityExpiringCache.get(host.getUrl());
+      if (availability != null) {
+        host.setAvailability(availability);
+      }
+    }
+  }
+
+  @Override
+  public void releaseResources() {
+    LOGGER.fine(() -> Messages.get("PluginServiceImpl.releaseResources"));
+
+    try {
+      if (this.currentConnection != null && !this.currentConnection.isClosed()) {
```

Comment:
We can't close current connection here. The whole idea of a "current connection" makes no sense for monitoring service and monitors.

Reply:
Wouldn't current connection just refer to the current monitoring connection in use? Or is this variable not intended to be used at all in monitoring connections? Should currentConnection from the class?